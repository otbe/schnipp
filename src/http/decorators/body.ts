import { createMethodParameterDecorator } from '../factories/createMethodParamDecorator';

export const Body = () =>
  createMethodParameterDecorator(e => JSON.parse(e.body || '{}'));
