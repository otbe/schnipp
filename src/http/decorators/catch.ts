import { inject } from 'simple-ts-di';

const CATCH = Symbol('catch');

export const Catch = (...exceptions: any[]): ClassDecorator => (
  target: any
) => {
  inject()(target);
  Reflect.defineMetadata(CATCH, exceptions, target.constructor);
};

export const getCaughtExceptions = (catcher: any) =>
  Reflect.getMetadata(CATCH, catcher.constructor) || [];
