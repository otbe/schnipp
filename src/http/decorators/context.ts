import { createMethodParameterDecorator } from '../factories/createMethodParamDecorator';

export const Context = () =>
  createMethodParameterDecorator((e, m, context) => context);
