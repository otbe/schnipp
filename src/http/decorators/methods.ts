import { createHttpMethodDecorator } from '../factories/createHttpMethodDecorator';
import { HttpVerb } from '../utils/AuthPolicy';

export const GET = createHttpMethodDecorator(HttpVerb.GET);
export const POST = createHttpMethodDecorator(HttpVerb.POST);
export const DELETE = createHttpMethodDecorator(HttpVerb.DELETE);
export const HEAD = createHttpMethodDecorator(HttpVerb.HEAD);
export const OPTIONS = createHttpMethodDecorator(HttpVerb.OPTIONS);
export const PATCH = createHttpMethodDecorator(HttpVerb.PATCH);
export const PUT = createHttpMethodDecorator(HttpVerb.PUT);
export const ALL = createHttpMethodDecorator(HttpVerb.ALL);
