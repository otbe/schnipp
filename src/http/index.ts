import {
  APIGatewayEvent,
  APIGatewayProxyResult,
  CustomAuthorizerEvent,
  CustomAuthorizerResult
} from 'aws-lambda';
import { IHandler } from '..';

export type APIGatewayHandler = IHandler<
  APIGatewayEvent,
  APIGatewayProxyResult
>;

export type AuthorizeHandler = IHandler<
  CustomAuthorizerEvent,
  CustomAuthorizerResult
>;

export { AuthPolicy, HttpVerb, Effect } from './utils/AuthPolicy';
export { createResponse } from './utils/createResponse';
export { MetaData } from './utils/MetaData';
export {
  DecoratedExceptionFilter,
  ExceptionFilter
} from './utils/ExceptionFilter';
export { Guard } from './utils/Guard';
export { DefaultExecutionContext } from './utils/ExecutionContext';
export { Catch } from './decorators/catch';
