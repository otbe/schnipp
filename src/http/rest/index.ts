export { RestController } from './RestController';

export {
  InjectedParam,
  MethodMetaData,
  ParamInjector
} from './factories/createHttpMethodDecorator';

export {
  GET,
  POST,
  DELETE,
  PUT,
  PATCH,
  OPTIONS,
  HEAD
} from './decorators/methods';

export { Path } from './decorators/path';
export { Param } from './decorators/param';
export { Query } from './decorators/query';
export { StatusCode } from './decorators/statusCode';
export { Context } from './decorators/context';
export { Body } from './decorators/body';
export { Header } from './decorators/header';
export { Meta } from './decorators/meta';
export { Use } from './decorators/use';

export { HttpException } from './exceptions/HttpException';
export { NotFoundException } from './exceptions/NotFoundException';
export { ForbiddenException } from './exceptions/ForbiddenException';
