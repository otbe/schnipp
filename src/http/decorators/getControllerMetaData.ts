import { MetaData, Guard, DecoratedExceptionFilter } from '..';
import { Newable } from 'simple-ts-di';
import { RestController } from '../rest';
import { GraphQLController } from '../graphql';
import { ResolverMeta } from '../graphql/decorators/method';

export const CONTROLLER = Symbol('controller');

export type ControllerBaseData = {
  headers: { [key: string]: string };
  filters: Array<DecoratedExceptionFilter>;
  guards: Array<Newable<Guard>>;
  metaData: MetaData;
};

export type GraphQLControllerData = ControllerBaseData & {
  resolvers?: Array<ResolverMeta>;
};

export type RestControllerData = ControllerBaseData & {
  basePath?: string;
};

export type ControllerData = GraphQLControllerData | RestControllerData;

export const getControllerMetaData = (
  controller: typeof RestController | typeof GraphQLController
) => {
  let data: ControllerData | null = Reflect.getMetadata(CONTROLLER, controller);

  if (data == null) {
    data = {
      filters: [],
      guards: [],
      metaData: {},
      headers: {}
    };

    Reflect.defineMetadata(CONTROLLER, data, controller);
  }

  return data;
};
