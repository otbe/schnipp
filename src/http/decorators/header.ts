import { getControllerMetaData } from './getControllerMetaData';
import { getMethodMetaData } from '../factories/createHttpMethodDecorator';

export const Header = (name: string, value: string) => (
  ...args: Array<any>
) => {
  const target = args[0];

  switch (args.length) {
    // class
    case 1:
      const controllerMetaData = getControllerMetaData(target);
      controllerMetaData.headers[name] = value;
      return;
    // method
    case 2:
    case 3:
      const methodMetaData = getMethodMetaData(args[1], target);
      methodMetaData.headers[name] = value;
      return args[2];
    default:
      throw 'not supported target for @Header';
  }
};
