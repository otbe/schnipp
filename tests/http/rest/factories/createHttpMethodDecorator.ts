import 'reflect-metadata';
import {
  getMethodMetaData,
  findMethod
} from '../../../../src/http/rest/factories/createHttpMethodDecorator';
import { Path, GET, POST } from '../../../../src/http/rest';
import { ALL } from '../../../../src/http/rest/decorators/methods';

describe('getMethodMetaData', () => {
  it('should return default meta data', () => {
    class Test {
      test() {}
    }

    const f = new Test();
    expect(getMethodMetaData('test', f)).toMatchSnapshot();
  });
});

describe('findMethod', () => {
  it('find method for request', () => {
    @Path('/events')
    class Test {
      @GET()
      get() {}

      @GET('/{id}')
      detail() {}

      @POST()
      post() {}

      @ALL()
      all() {}
    }

    const f = new Test();
    expect(findMethod('/events', '/events', 'GET', f)).toMatchSnapshot();
    expect(findMethod('/events', '/events/{id}', 'GET', f)).toMatchSnapshot();
    expect(findMethod('/events', '/events', 'POST', f)).toMatchSnapshot();
    expect(
      findMethod('/events', '/events/fooo', 'OPTIONS', f)
    ).toMatchSnapshot();
  });
});
