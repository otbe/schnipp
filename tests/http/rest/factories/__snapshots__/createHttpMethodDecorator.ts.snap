// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findMethod find method for request 1`] = `
Object {
  "filters": Array [],
  "guards": Array [],
  "headers": Object {},
  "metaData": Object {},
  "method": "GET",
  "paramaters": Array [],
  "path": "/",
  "propertyKey": "get",
  "statusCode": 200,
}
`;

exports[`findMethod find method for request 2`] = `
Object {
  "filters": Array [],
  "guards": Array [],
  "headers": Object {},
  "metaData": Object {},
  "method": "GET",
  "paramaters": Array [],
  "path": "/{id}",
  "propertyKey": "detail",
  "statusCode": 200,
}
`;

exports[`findMethod find method for request 3`] = `
Object {
  "filters": Array [],
  "guards": Array [],
  "headers": Object {},
  "metaData": Object {},
  "method": "POST",
  "paramaters": Array [],
  "path": "/",
  "propertyKey": "post",
  "statusCode": 201,
}
`;

exports[`findMethod find method for request 4`] = `
Object {
  "filters": Array [],
  "guards": Array [],
  "headers": Object {},
  "metaData": Object {},
  "method": "*",
  "paramaters": Array [],
  "path": "/",
  "propertyKey": "all",
  "statusCode": 200,
}
`;

exports[`getMethodMetaData should return default meta data 1`] = `
Object {
  "filters": Array [],
  "guards": Array [],
  "headers": Object {},
  "metaData": Object {},
  "paramaters": Array [],
  "path": "/",
  "propertyKey": "test",
}
`;
